go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,484 @@
+		writer = gzip
+	}
+
+	if err = backupCmd.Start(); err != nil {
+		return false, vterrors.Wrap(err, "unable to start backup")
+	}
+
+	// Copy from the stream output to destination file (optional gzip)
+	_, err = io.Copy(writer, backupOut)
+	if err != nil {
+		return false, vterrors.Wrap(err, "cannot copy output from xtrabackup command")
+	}
+
+	// Close gzip to flush it, after that all data is sent to writer.
+	if gzip != nil {
+		if err = gzip.Close(); err != nil {
+			return false, vterrors.Wrap(err, "cannot close gzip")
+		}
+	}
+
+	// Flush the buffer to finish writing on destination.
+	if err = dst.Flush(); err != nil {
+		return false, vterrors.Wrapf(err, "cannot flush destination: %v", backupFileName)
+	}
+
+	errOutput, err := ioutil.ReadAll(backupErr)
+	backupCmd.Wait()
+	output := string(errOutput)
+
+	logger.Infof("Xtrabackup backup command output: %v", output)
+	// check for success message : xtrabackup: completed OK!
+	usable := (err == nil && strings.Contains(output, successMsg))
+	substrs := strings.Split(output, "'")
+	index := -1
+	for i, str := range substrs {
+		if strings.Contains(str, "GTID of the last change") {
+			index = i + 1
+			break
+		}
+	}
+	position := ""
+	if index != -1 {
+		// since we are extracting this from the log, it contains newlines
+		// replace them with a single space to match the SET GLOBAL gtid_purged command in xtrabackup_slave_info
+		position = strings.Replace(substrs[index], "\n", " ", -1)
+	}
+	logger.Infof("Found position: %v", position)
+
+	// flavor is required to parse a string into a mysql.Position
+	var replicationPosition mysql.Position
+	if replicationPosition, err = mysql.ParsePosition(flavor, position); err != nil {
+		return false, err
+	}
+
+	// open the MANIFEST
+	mwc, err := bh.AddFile(ctx, backupManifest, 0)
+	if err != nil {
+		return usable, vterrors.Wrapf(err, "cannot add %v to backup", backupManifest)
+	}
+	defer closeFile(mwc, backupManifest)
+
+	// JSON-encode and write the MANIFEST
+	bm := &XtraBackupManifest{
+		FileName:     backupFileName,
+		BackupMethod: xtrabackup,
+		Position:     replicationPosition,
+		SkipCompress: !*backupStorageCompress,
+		Params:       *xtrabackupBackupFlags,
+	}
+
+	data, err := json.MarshalIndent(bm, "", "  ")
+	if err != nil {
+		return usable, vterrors.Wrapf(err, "cannot JSON encode %v", backupManifest)
+	}
+	if _, err := mwc.Write([]byte(data)); err != nil {
+		return usable, vterrors.Wrapf(err, "cannot write %v", backupManifest)
+	}
+
+	return usable, nil
+}
+
+// ExecuteRestore restores from a backup. Any error is returned.
+func (be *XtrabackupEngine) ExecuteRestore(
+	ctx context.Context,
+	cnf *Mycnf,
+	mysqld MysqlDaemon,
+	logger logutil.Logger,
+	dir string,
+	bhs []backupstorage.BackupHandle,
+	restoreConcurrency int,
+	hookExtraEnv map[string]string) (mysql.Position, error) {
+
+	var bh backupstorage.BackupHandle
+	var bm XtraBackupManifest
+	var toRestore int
```

Comment:
what units is this var? can you name it something more specific?

filesToRestore? records? something else?

Reply:
this is just an index to loop through the list of backups. renamed it to index.